### PR TITLE
feat: manage ls binary automatically [ROAD-1054]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

Automatically download & upload LS if feature flag is enabled and automatic management of binaries is enabled. This uses the same logic as the CLI download. If feature flag is enabled, CLI download is deactivated, as LS will take care of it.

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: src/snyk/cli/api/staticCliApi.ts [hunk 1]
```diff
@@ -4,12 +4,11 @@ import { getAxiosProxyConfig } from '../../common/proxy';
 import { IVSCodeWorkspace } from '../../common/vscode/workspace';
 import { CliExecutable } from '../cliExecutable';
 import { CliSupportedPlatform } from '../supportedPlatforms';
-
-export type CliDownloadAxiosResponse = { data: stream.Readable; headers: { [header: string]: unknown } };
+import { DownloadAxiosResponse } from '../../common/download/downloader';
```
Comment (bastiandoetsch) [inline src/snyk/cli/api/staticCliApi.ts:7]: optimize imports

File: src/snyk/base/modules/snykLib.ts [hunk 2]
```diff
@@ -127,7 +127,7 @@ export default class SnykLib extends BaseSnykModule implements ISnykLib {
     if (!this.ossService) throw new Error('OSS service is not initialized.');
 
     // wait until Snyk CLI is downloaded
-    await firstValueFrom(this.cliDownloadService.cliIsReady$);
+    await firstValueFrom(this.downloadService.downloadReady);
```
Comment (bastiandoetsch) [inline src/snyk/base/modules/snykLib.ts:130]: add check before starting LS, too

File: src/snyk/common/languageServer/languageServer.ts [hunk 3]
```diff
@@ -93,6 +93,10 @@ export class LanguageServer implements ILanguageServer {
       });
     });
 
+    this.client.onNotification(SNYK_CLI_PATH, ({ cliPath }: { cliPath: string }) => {
+      void this.configuration.setCliPath(cliPath);
```
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/languageServer.ts:97]: catch & error handling like above

File: src/snyk/common/configuration/configuration.ts [hunk 4]
```diff
@@ -213,10 +218,15 @@ export class Configuration implements IConfiguration {
   }
 
   async setToken(token: string | undefined): Promise<void> {
-    if (!token) return;
+    if (!token || token === '') return;
```
Comment (bastiandoetsch) [inline src/snyk/common/configuration/configuration.ts:221]: remove `===`

File: src/snyk/common/download/downloader.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    private readonly extensionDir: string,
+    private readonly window: IVSCodeWindow,
+    private readonly logger: ILog,
+  ) {}
+
+  /**
+   * Downloads CLI or LS to the extension folder. Existing executable is deleted.
+   */
+  async download(): Promise<CliExecutable | LsExecutable | null> {
+    if (!this.configuration.getPreviewFeatures().lsAuthenticate) {
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:36]: add todo to remove feature flag when not needed anymore

File: src/snyk/common/download/downloader.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    }
+  }
+
+  private async getLsExecutable(lsPlatform: LsSupportedPlatform): Promise<LsExecutable | null> {
+    const lsPath = LsExecutable.getPath(this.configuration.getSnykLanguageServerPath());
+    if (await this.binaryExists(lsPath)) {
+      await this.deleteFileAtPath(lsPath);
+    }
+
+    const lsVersion = await this.lsApi.getMetadata().then(metadata => metadata.version);
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:58]: `(await this.lsApi.getMetadata()).version`

File: src/snyk/common/languageServer/staticLsApi.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  getSha256Checksum(platform: LsSupportedPlatform): Promise<string>;
+}
+
+export class StaticLsApi implements IStaticLsApi {
+  private readonly baseUrl = `https://static.snyk.io/snyk-ls/${PROTOCOL_VERSION}`;
+
+  constructor(private readonly workspace: IVSCodeWorkspace) {}
+
+  async getDownloadUrl(platform: LsSupportedPlatform): Promise<string> {
+    return `${this.baseUrl}/${await this.getFileName(platform)}`;
```
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/staticLsApi.ts:35]: `getVersionedFilepath`

File: src/snyk/common/languageServer/staticLsApi.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  async getFileName(platform: LsSupportedPlatform): Promise<string> {
```
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/staticLsApi.ts:38]: combine with function above

File: src/snyk/common/languageServer/staticLsApi.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        checksum = line.split(' ')[0].trim().toLowerCase();
+      }
+    });
+    if (checksum == '') return Promise.reject(new Error('Checksum not found'));
+
+    return checksum;
+  }
+
+  async getMetadata(): Promise<LsMetadata> {
+    const response = await axios.get<string>(`${this.baseUrl}/metadata.json`, getAxiosProxyConfig(this.workspace));
```
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/staticLsApi.ts:78]: `AxiosResponse<LsMetadata>`

File: src/snyk/common/platform.ts [hunk 10]
```diff
@@ -1,10 +1,27 @@
 import * as os from 'os';
+import { LsSupportedPlatform, SupportedLsPlatformsList } from './languageServer/supportedPlatforms';
 
 export class Platform {
   static getCurrent(): NodeJS.Platform {
     return os.platform();
   }
 
+  static getCurrentWithArch(): LsSupportedPlatform | null {
```
Comment (bastiandoetsch) [inline src/snyk/common/platform.ts:9]: check for usage with jason's pr

File: src/snyk/extension.ts [hunk 11]
```diff
@@ -320,6 +321,8 @@ class SnykExtension extends SnykLib implements IExtension {
       configuration,
     );
 
+    await this.languageServer.start();
```
Comment (bastiandoetsch) [inline src/snyk/extension.ts:324]: wait for downloadReady

File: src/test/integration/analytics.test.ts [hunk 12]
```diff
@@ -9,7 +9,7 @@ suite('Analytics', () => {
   let welcomeIsViewed: sinon.SinonSpy;
 
   setup(async () => {
-    await configuration.setToken(''); // ensures user is not authenticated
+    await configuration.setToken(undefined); // ensures user is not authenticated
```
Comment (bastiandoetsch) [inline src/test/integration/analytics.test.ts:12]: clearToken()

File: src/test/integration/viewNavigation.test.ts [hunk 13]
```diff
@@ -8,12 +8,12 @@ import { configuration } from '../../snyk/common/configuration/instance';
 
 suite('View Navigation', () => {
   setup(async () => {
-    await configuration.setToken(undefined);
+    await configuration.setToken('');
```
Comment (bastiandoetsch) [inline src/test/integration/viewNavigation.test.ts:11]: clearToken()

File: src/test/integration/viewNavigation.test.ts [hunk 14]
```diff
@@ -8,12 +8,12 @@ import { configuration } from '../../snyk/common/configuration/instance';
...
     await configuration.setFeaturesConfiguration(undefined);
   });
 
   teardown(async () => {
-    await configuration.setToken(undefined);
+    await configuration.setToken('');
```
Comment (bastiandoetsch) [inline src/test/integration/viewNavigation.test.ts:16]: clearToken()

File: src/test/unit/download/downloader.test.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    sinon.restore();
+  });
+
+  test('Download of LS fails if platform is not supported', async () => {
+    const downloader = new Downloader(configuration, cliApi, lsApi, extensionDir, windowMock, logger);
+    sinon.stub(Platform, 'getCurrentWithArch').returns(null);
+    await rejects(() => downloader.download());
+  });
+
+  test('Download of LS  removes executable, if it exists', async () => {
```
Comment (bastiandoetsch) [inline src/test/unit/download/downloader.test.ts:143]: remove redundant space

File: src/test/unit/common/services/downloadService.test.ts [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+  test("Doesn't update LS if 3 days passed since last update", async () => {
+    const service = new DownloadService(context, configuration, cliApi, lsApi, windowMock, logger, downloader);
+    stub(service, 'isCliInstalled').resolves(true);
+    stub(service, 'isLsInstalled').resolves(true);
+
+    const threeDaysInMs = 3 * 24 * 3600 * 1000;
+    contextGetGlobalStateValue.withArgs(MEMENTO_CLI_LAST_UPDATE_DATE).returns(Date.now() - threeDaysInMs);
+    contextGetGlobalStateValue.withArgs(MEMENTO_LS_LAST_UPDATE_DATE).returns(Date.now() - threeDaysInMs);
+
+    sinon.stub(downloader, 'download').resolves(new CliExecutable('1.0.0', new Checksum('test')));
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:300]: Fix this

File: src/test/unit/common/services/downloadService.test.ts [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+      getPreviewFeatures: () => {
+        return {
+          lsAuthenticate: true,
+        };
+      },
+      getSnykLanguageServerPath: () => 'abc/d',
+    } as unknown as IConfiguration;
+    const service = new DownloadService(context, configuration, cliApi, lsApi, windowMock, logger, downloader);
+    contextGetGlobalStateValue.withArgs(MEMENTO_CLI_CHECKSUM).returns(undefined);
+    contextGetGlobalStateValue.withArgs(MEMENTO_CLI_LAST_UPDATE_DATE).returns(undefined);
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:336]: fix me

File: src/snyk/common/services/downloadService.ts [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+import { Downloader } from '../download/downloader';
+import { messages } from '../../cli/messages/messages';
+import { CliSupportedPlatform, isPlatformSupported } from '../../cli/supportedPlatforms';
+import { LsExecutable } from '../languageServer/lsExecutable';
+import { IStaticLsApi } from '../languageServer/staticLsApi';
+import { LsSupportedPlatform } from '../languageServer/supportedPlatforms';
+
+export class DownloadService {
+  readonly fourDaysInMs = 4 * 24 * 3600 * 1000;
+  readonly downloadReady = new ReplaySubject<void>(1);
```
Comment (michelkaporin) [inline src/snyk/common/services/downloadService.ts:25]: Observables are marked with dollar sign `$` in JavaScript

File: src/snyk/common/languageServer/lsExecutable.ts [hunk 19]
```diff
@@ -2,33 +2,41 @@ import os from 'os';
 import path from 'path';
 import { Checksum } from '../../cli/checksum';
 import { Platform } from '../platform';
-import { LsSupportedPlatform } from './supportedPlatforms';
+import { LsSupportedPlatform, SupportedLsPlatformsList } from './supportedPlatforms';
+import fs from 'fs/promises';
+import { IConfiguration } from '../configuration/configuration';
+import { mkdir, mkdirSync } from 'fs';
```
Comment (michelkaporin) [inline src/snyk/common/languageServer/lsExecutable.ts:8]: Let's fix the warnings

**Pull request merged**